src/capstone.rs

```diff
@@ -1,63 +1,146 @@
+    /// ```
+    pub fn new() -> CapstoneBuilder {
+        CapstoneBuilder::new()
+    }
+
+    /// Create a new instance of the decompiler using the "raw" interface.
+    /// The user must ensure that only sensical `Arch`/`Mode` combinations are used.
     ///
     /// ```
-    /// use capstone::{Arch, Capstone, Mode};
-    /// let cs = Capstone::new(Arch::X86, Mode::Mode64);
+    /// use capstone::{Arch, Capstone, NO_EXTRA_MODE, Mode};
+    /// let cs = Capstone::new_raw(Arch::X86, Mode::Mode64, NO_EXTRA_MODE, None);
     /// assert!(cs.is_ok());
     /// ```
-    pub fn new(arch: Arch, mode: Mode) -> CsResult<Capstone> {
+    pub fn new_raw<T: Iterator<Item = ExtraMode>>(
+        arch: Arch,
+        mode: Mode,
+        extra_mode: T,
+        endian: Option<Endian>,
+    ) -> CsResult<Capstone> {
         let mut handle = 0;
         let csarch: cs_arch = arch.into();
         let csmode: cs_mode = mode.into();
-        let err = unsafe { cs_open(csarch, csmode, &mut handle) };
 
-        if cs_err::CS_ERR_OK == err {
-            let mut opt_state: HashMap<cs_opt_type, libc::size_t> = HashMap::new();
-            opt_state.insert(cs_opt_type::CS_OPT_SYNTAX,
-                             CS_OPT_SYNTAX_DEFAULT as libc::size_t);
-            opt_state.insert(cs_opt_type::CS_OPT_DETAIL,
-                             cs_opt_value::CS_OPT_OFF as libc::size_t);
-            opt_state.insert(cs_opt_type::CS_OPT_MODE, mode as libc::size_t);
-            opt_state.insert(cs_opt_type::CS_OPT_MEM, 0);
-            opt_state.insert(cs_opt_type::CS_OPT_SKIPDATA,
-                             cs_opt_value::CS_OPT_OFF as libc::size_t);
-
-            Ok(Capstone {
-                   csh: handle,
-                   cs_option_state: opt_state,
-                   _arch: arch,
-               })
-        } else {
-            Err(err.into())
-        }
-    }
+        let endian = match endian {
+            Some(endian) => cs_mode::from(endian) as usize,
+            None => 0,
+        };
+        let extra_mode = Self::extra_mode_value(extra_mode);
+
+        let err = unsafe {
+            let combined_mode: usize = csmode as usize | endian | extra_mode;
+            let combined_mode: cs_mode = mem::transmute(combined_mode as CsModeRepr);
+            cs_open(csarch, combined_mode, &mut handle)
+        };
 
-    #[inline]
-    fn set_option(&self, opt_type: cs_opt_type, value: usize) -> CsResult<()> {
-        let err = unsafe { cs_option(self.csh, opt_type, value) };
         if cs_err::CS_ERR_OK == err {
-            Ok(())
+            let syntax = CS_OPT_SYNTAX_DEFAULT as usize;
+            let raw_mode = 0usize;
+            let detail_enabled = false;
+
+            let mut cs = Capstone {
+                csh: handle,
+                syntax,
+                endian,
+                mode: csmode as usize,
+                extra_mode,
+                detail_enabled,
+                raw_mode,
+                _arch: arch,
```

Comment:
why is this `_arch` and not just `arch`?

Reply:
Actually, according to `git blame`, you introduced `_arch` in b3869291.
=)

Reply:
:sob: